transformer_engine/pytorch/transformer.py

```diff
@@ -219,6 +228,243 @@ def forward(
         return context_layer
 
 
+class FlashAttention(torch.nn.Module):
+    """Dot product attention implementation by using the flash-attn package.
+    """
+
+    def __init__(
+        self,
+        num_attention_heads: int,
+        kv_channels: int,
+        attention_dropout: float = 0.0,
+        layer_number: Optional[int] = None,
+        apply_query_key_layer_scaling: bool = True,
+        attention_softmax_in_fp32: bool = True,
+        tp_size: int = 1,
+        get_rng_state_tracker: Optional[Callable] = None,
+        sequence_parallel: bool = False,
+        attn_mask_type: str = "causal",
+    ) -> None:
+        super().__init__()
+
+        if "hopper" not in _flash_attn_version:
+            raise ImportError(
+                'Please install correct version of flash-attn with ' \
+                'pip install git+https://github.com/ksivaman/flash-attention.git@hopper. ' \
+                'If running on Hopper, ' \
+                'please install from source with compute capability 9.0.')
+        assert (
+            attn_mask_type == "causal"
+            ), 'FlashAttention currently only supports causal attention mask.'
+        assert (
+            attention_softmax_in_fp32 is True
+            ), 'FlashAttention currently only supports softmax compute in fp32.'
+
+        self.attn_causal_mask = attn_mask_type == "causal"
+        self.apply_query_key_layer_scaling = apply_query_key_layer_scaling
+
+        if layer_number is None:
+            self.apply_query_key_layer_scaling = False
+        else:
+            self.layer_number = max(1, layer_number)
+
+        projection_size = kv_channels * num_attention_heads
```

Comment:
A lot of that logic is the same as the unfused attention - why not have that in the actual DotProductAttention class instead of repeated code here?